### ISSUE TITLE
Using cleanlab to find no-iid, outlier, near_duplicate is time-consuming with high-dimesional features

### ISSUE THREAD

**Issue by xiaokening** [cleanlab/cleanlab#780]
the shape of pred_probs is (351133, 1522), the shape of embedding is (351133, 45812)

the code snippet is below：
`lab = Datalab(data_for_train, label_name="vps")`
`lab.find_issues(pred_probs=pred_probs,features=embeddings,issue_types={"outlier": {}, "near_duplicate": {}, "non_iid": {}})`
 `lab.report()`

This code has been running for 67 hours, but it is not over yet. Is this normal?

**Comment by jwmueller** [comment]
@xiaokening That seems overly slow, could you help us narrow down which part is the slow bit?

Note you can run find_issues() incrementally to check for one issue type at a time, so you can easily narrow down which type is the bottleneck for your dataset.

If you can share any runtime information and other details about your data, that would be really helpful!

**Comment by jwmueller** [comment]
For larger datasets, you may also consider trying [Cleanlab Studio](https://cleanlab.ai) which can run them more seamlessly on fully managed compute infrastructure

**Comment by xiaokening** [comment]
@jwmueller Thanks for your reply, I will try your proposed method later.

Below are the details of my data:
data_for_train(pd.DataFrame): the dateset for multilabel-classification algorithm;

pred_probs(nd.array): this matrix is converted from the csr matrix(by to_array() method) and represents the predicted label   
probability, the shape is  (351133, 1522);

embedding(nd.array):this matrix is converted from the csr matrix(by to_array() method) and which is composed of the 
pooler_output(351133, 1024) of bert model  and the tfidf feature(351133, 45684),  the embeddings shape is (351133, 46708);

**Comment by jwmueller** [comment]
Thanks! If you could share the runtimes of each issue check on this data, that would be very helpful for us to improve things. 

Beyond skipping the issue check that is overly slow, you can get easy speed improvements as follows:

1. Find label issues like this instead of using Datalab:
https://docs.cleanlab.ai/stable/tutorials/faq.html#How-can-I-find-label-issues-in-big-datasets-with-limited-memory?

2. Apply dimension-reduction to your embeddings before you feed them into Datalab. I suggest applying PCA to reduce the TF-IDF features before you concat them with the Bert features, cf:
https://www.kaggle.com/code/albeffe/text-clustering-tfidf-pca-beginner-tutorial

**Comment by xiaokening** [comment]
@jwmueller I'm not currently able to get the runtimes for each issue check on my data， I will share it with you later.

But now I have another question here.

is the runtimes for each issue check when I run find_issues() incrementally and the runtimes for each issue check when I run find_issues() simultaneously the same?

Looking forward to your reply, thanks!

**Comment by jwmueller** [comment]
> is the runtimes for each issue check when I run find_issues() incrementally and the runtimes for each issue check when I run find_issues() simultaneously the same?

Yes it should be

**Comment by elisno** [comment]
Hi @xiaokening, I think there are a few points I can help clarify.


---

> `lab.find_issues(pred_probs=pred_probs,features=embeddings,issue_types={"outlier": {}, "near_duplicate": {}, "non_iid": {}})`


All of the `issue_types` you've specified will use the `embeddings` and not the `pred_probs`, so we can set that aside for this discussion.

> the shape of embedding is (351133, 45812)

>This code has been running for 67 hours, but it is not over yet. Is this normal?

- The bottleneck is undoubtedly the nearest neighbor search within the `embeddings`.
- The k-nearest neighbors will be stored for subsequent issue checks, provided that the value of `k` is not larger than the current number of neighbors found.
- I see you're running the issue checks with the default hyperparameters.
  - The outliers check will require more neighbors than the others, meaning that the neighbor search will only need to be run once.

---

> Apply dimensionality reduction to your embeddings before you feed them into Datalab. I suggest applying PCA to reduce the TF-IDF features before you concatenate them with the Bert features

I agree with @jwmueller that reducing the dimensionality of your problem could help.

When concatenating BERT embeddings with TF-IDF features, consider these two aspects:

- Feature Scaling: BERT and TF-IDF features can have different scales, which could distort distance computations. You might want to apply feature scaling to ensure all features contribute equally. Here's a guide on feature scaling:
https://www.kaggle.com/code/discdiver/guide-to-scaling-and-standardizing

- Sparsity: TF-IDF features are typically sparse and maintaining this sparsity can save substantial computational resources. I'd suggest looking into sparsity-aware methods for handling these features, especially when dealing with large datasets. Maybe this matters less if you've applied dimensionality reduction to the features.

**Comment by xiaokening** [comment]
@jwmueller  Below is the runtimes for each issue check on my data.

near_duplicate: 27 hours and 20 minutes；
non_iid_issue: 89 hours and 24 minutes；
outlier_issue:  90 hours and 17 minutes；


@elisno Before concating BERT and TF-IDF features, I have scaled the bert feature with the sklearn.preprocessing.normalize function.

According to you, sparsity-aware algorithms can indeed speed up finding problems with data, I will try it later.

Thanks for your clarify.

**Comment by jwmueller** [comment]
@xiaokening Thanks for sharing the runtimes!  Could you share the code you used to produce these?  I'm wondering whether this is from a non-incremental run with separate Datalab for each issue-type, or if you ran these 3 issue checks incrementally from within the same Datalab instance.

**Comment by xiaokening** [comment]
@jwmueller  Below is the code.
```
import os
import logging
import warnings

import numpy as np
import pandas as pd

from pecos.utils import cli, logging_util, smat_util, torch_util
from cleanlab import Datalab
from cleanlab.multilabel_classification.filter import find_label_issues
from cleanlab.internal.multilabel_utils import onehot2int, int2onehot
from cleanlab.multilabel_classification import get_label_quality_scores

LOGGER = logging.getLogger(__name__)
warnings.simplefilter('ignore', UserWarning)
base_dir = os.getcwd()


def test_cleanlab():
    data_for_train = pd.read_excel(os.path.join(base_dir, "datasets/data_for_train.xlsx"))
    LOGGER.info(f"loaded data_for_train.xlsx, total {len(data_for_train)} samples")
    
    pred_probs = smat_util.load_matrix("/home/models/ch_nps_viewpoint_model/chinese-macbert-large/pred_probs.npz",
                                       dtype=np.float32).toarray()
    LOGGER.info("Loaded pred_probs matrix with shape={}".format(pred_probs.shape))
    
    embeddings = smat_util.load_matrix("/home/models/ch_nps_viewpoint_model/chinese-macbert-large/embeddings.npz",
                                       dtype=np.float32).toarray()
    LOGGER.info("Loaded embeddings matrix with shape={}".format(embeddings.shape))
    
    #cleanlab查找outlier, near_duplicate, non_iid
    lab = Datalab(data_for_train, label_name="vps", verbosity=3) # label_name没有用
    
    lab.find_issues(
        pred_probs=pred_probs,
        features=embeddings,
        issue_types={"outlier": {}}
    )
    lab.report()
    
    outlier_issue = lab.get_issues("outlier")
    data_for_train_outlier = pd.concat([data_for_train, outlier_issue], axis=1)
    
    data_for_train_outlier.to_excel("/home/data_for_train_outlier.xlsx", index=False)
    tgt_dir = os.path.join(base_dir, "datasets/data_for_train_outlier.xlsx")
    os.system(f"mv /home/data_for_train_outlier.xlsx {tgt_dir}")
    
    LOGGER.info(f"use cleanlab to find outlier issue completed! saved to {tgt_dir} completed")
    
    lab.find_issues(
        pred_probs=pred_probs,
        features=embeddings,
        issue_types={"near_duplicate": {}}
    )
    lab.report()
    
    near_duplicate = lab.get_issues("near_duplicate")
    data_for_train_near_duplicate = pd.concat([data_for_train, near_duplicate], axis=1)
    
    data_for_train_near_duplicate.to_excel("/home/data_for_train_near_duplicate.xlsx", index=False)
    tgt_dir = os.path.join(base_dir, "datasets/data_for_train_near_duplicate.xlsx")
    os.system(f"mv /home/data_for_train_near_duplicate.xlsx {tgt_dir}")
    
    LOGGER.info(f"use cleanlab to find near_duplicate issue completed! saved to {tgt_dir} completed")
    
    lab.find_issues(
        pred_probs=pred_probs,
        features=embeddings,
        issue_types={"non_iid": {}}
    )
    lab.report()
    
    non_iid_issue = lab.get_issues("non_iid")
    data_for_train_non_iid_issue = pd.concat([data_for_train, non_iid_issue], axis=1)
    
    data_for_train_non_iid_issue.to_excel("/home/data_for_train_non_iid_issue.xlsx", index=False)
    tgt_dir = os.path.join(base_dir, "datasets/data_for_train_non_iid_issue.xlsx")
    os.system(f"mv /home/data_for_train_non_iid_issue.xlsx {tgt_dir}")
    
    LOGGER.info(f"use cleanlab to non_iid issue completed! saved to {tgt_dir} completed")
if __name__ == "__main__":
    logging_util.setup_logging_config(level=2)
    test_cleanlab()
```

Every time I check out one issue type, I comment out the code snippets for the other two issue type. Therefore, I run a non-incremental check with separate Datalab for each issue-type.  I check out  three issue types simultaneously！

**Comment by xiaokening** [comment]
@jwmueller  After I use cleanlab to find near_duplicate_issue  of my data, I don't know how to deal with the data whose is_near_duplicate_issue is True(these data are indeed near duplicated) to improve the robustness of my model.

Can I just delete the data whose is_near_duplicate_issue is True ？ will it cause all near_duplicate data deleted ?  And not keep one of them?

**Comment by elisno** [comment]
@xiaokening Based on the size of your dataset, I think you would benefit more from Datalab by computing a knn-graph yourself.

Currently most of the `features`-based checks end up computing a knn-graph and reusing that.

If you can get away with it, it should be faster to run:

```python
datalab.find_issues(knn_graph=knn_graph, ...)
```

instead of 


```python
datalab.find_issues(features=X, ...)
```

If you provide both `features` and `knn_graph`, most of the relevant issue checks try to use the `knn_graph` if possible.

---

## How do I make a `knn_graph` that `Datalab` accepts?

Here's a simple function that turns arrays of `distances` and `indices` into a sparse matrix in CSR format, compatible with the `knn_graph` argument that Datalab accepts.

```python
def create_knn_graph(distances, indices):
    """
    Create a K-nearest neighbors (KNN) graph in CSR format from provided distances and indices.

    Parameters:
    distances (np.array): 2D array of shape (n_samples, n_neighbors) containing distances to nearest neighbors.
    indices (np.array): 2D array of shape (n_samples, n_neighbors) containing indices of nearest neighbors.

    Returns:
    scipy.sparse.csr_matrix: KNN graph in CSR format.
    """

    # Verify the input
    assert distances.shape == indices.shape, "distances and indices must have the same shape"

    n_samples, n_neighbors = distances.shape

    # Convert the indices and distances arrays to 1D
    indices_1d = indices.ravel()
    distances_1d = distances.ravel()

    # Construct a row index array
    row_indices = np.repeat(np.arange(n_samples), n_neighbors)

    # Create the CSR matrix
    return csr_matrix((distances_1d, (row_indices, indices_1d)), shape=(n_samples, n_samples))
```

There are several popular ways of computing distances between points.

Most people are familiar with `NearestNeighbors` class in the "scikit-learn" library.

A common workflow looks something like this:

```python
from sklearn.neighbors import NearestNeighbors

# Instantiate the NearestNeighbors class
nn = NearestNeighbors(n_neighbors=k, metric='cosine')

# Fit the model to the feature matrix, X
nn.fit(X)

# Get the distances and indices of the k nearest neighbors of each point
distances, indices = nn.kneighbors()

# Create the KNN graph
knn_graph = create_knn_graph(distances, indices)
```

---

### Other algorithms

Now, `NearestNeighbors` from scikit-learn may not be the fastest way to compute the distances.

Another common library for computing distances is "faiss". It requires some considerations to query the correct number of neighbors, and special care must be taken to the chosen metric returns distances instead of similarities.

```python
import faiss

# Faiss uses single precision, so we need to convert the data type
X_faiss = np.float32(X)

# Normalize the vectors
faiss.normalize_L2(X_faiss)

# Build the index
index = faiss.IndexFlatIP(X_faiss.shape[1])
index.add(X_faiss)

# Perform the search. This particular search includes the query points as well.
D, I = index.search(X_faiss, k+1)

# Remove the first column, which contains the distances to the query points themselves
D, I = D[:, 1:], I[:, 1:]

# Convert cosine similarity to cosine distance
D = 1 - D

# D contains the distances, and I contains the indices of the neighbors
knn_graph_faiss = create_knn_graph(D, I)
```


---

For now, if you're in doubt, make sure your `distances` and `indices` are compatible with those shown in the scikit-learn example.
